x/mongo/driver/connstring/connstring.go

```diff
@@ -137,7 +138,7 @@ func (u *ConnString) String() string {
 func (u *ConnString) HasAuthParameters() bool {
 	// Check all auth parameters except for AuthSource because an auth source without other credentials is semantically
 	// valid and must not be interpreted as a request for authentication.
-	return u.AuthMechanism != "" || u.AuthMechanismProperties != nil || u.Username != "" || u.PasswordSet
+	return u.AuthMechanism != "" || u.AuthMechanismProperties != nil || u.UsernameSet || u.PasswordSet
```

Comment:
`AuthMechanismPropertiesSet`: If `authMechanismProperties` is present in the URI, we make a map, so it would be non-nil, even if it were empty. Therefore, having the `nil` check is sufficient for determining if it was present or not.

`AuthMechanismSet`: I'm not 100% sure why this doesn't exist. For this function specifically, we only care if it's non-empty because there's no empty auth mechanism.

I agree the `X`/`XSet` method for tracking URI options is confusing, but I don't want to make any changes to that in here. Instead, that should be handled in https://jira.mongodb.org/browse/GODRIVER-1498.